votor/src/timer_manager/timers.rs

```diff
@@ -0,0 +1,243 @@
+        delta_timeout: Duration,
+        delta_block: Duration,
+        event_sender: Sender<VotorEvent>,
+    ) -> Self {
+        Self {
+            delta_timeout,
+            delta_block,
+            timers: HashMap::new(),
+            heap: BinaryHeap::new(),
+            event_sender,
+        }
+    }
+
+    /// Call to set timeouts for a new window of slots.
+    pub(super) fn set_timeouts(&mut self, slot: Slot) {
+        assert_eq!(self.heap.len(), self.timers.len());
+        let (timer, next_fire) = TimerState::new(slot, self.delta_timeout);
+        // It is possible that this slot already has a timer set e.g. if there
+        // are multiple ParentReady for the same slot.  Do not insert new timer then.
+        self.timers.entry(slot).or_insert_with(|| {
+            self.heap.push(Reverse((next_fire, slot)));
+            timer
+        });
+    }
+
+    /// Call to make progress on the timer states.  If there are still active
+    /// timer states, returns when the earliest one might become ready.
+    pub(super) fn progress(&mut self) -> Option<Instant> {
+        assert_eq!(self.heap.len(), self.timers.len());
+        let mut ret_timeout = None;
+        let now = Instant::now();
+        loop {
+            assert_eq!(self.heap.len(), self.timers.len());
+            match self.heap.pop() {
+                None => break,
+                Some(Reverse((next_fire, slot))) => {
+                    if now < next_fire {
+                        ret_timeout = Some(match ret_timeout {
+                            None => next_fire,
+                            Some(r) => std::cmp::min(r, next_fire),
+                        });
+                        self.heap.push(Reverse((next_fire, slot)));
+                        break;
+                    }
+
+                    let mut timer = self.timers.remove(&slot).unwrap();
+                    if let Some(event) = timer.progress(self.delta_block) {
+                        self.event_sender.send(event).unwrap();
+                    }
+                    if let Some(next_fire) = timer.next_fire() {
+                        self.heap.push(Reverse((next_fire, slot)));
+                        assert!(self.timers.insert(slot, timer).is_none());
+                        ret_timeout = Some(match ret_timeout {
+                            None => next_fire,
+                            Some(r) => std::cmp::min(r, next_fire),
+                        });
+                    }
+                }
+            }
+        }
+        ret_timeout
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, crossbeam_channel::unbounded, std::thread};
+
+    #[test]
+    fn timer_state_machine() {
+        let slot = 0;
+        let (mut timer_state, next_fire) = TimerState::new(slot, Duration::from_micros(1));
+
+        let next_fire = next_fire.duration_since(Instant::now());
+        let delta_block = Duration::from_micros(1);
+
+        thread::sleep(next_fire);
```

Comment:
What do you think about having timer_state.progress and timers.progress take in an Instant that we pass in from the TimerManager. Then we can avoid having to sleep in test code